External/Plugins/HaXeContext/Linters/DiagnosticsLinter.cs

```diff
@@ -1,121 +1,216 @@
+                    };
+
+                    switch (res.Severity)
                     {
-                        PluginBase.RunAsync(() =>
-                        {
-                            TraceManager.Add(hc.Errors, (int) TraceType.Error);
-                        });
+                        case HaxeDiagnosticsSeverity.INFO:
+                            result.Severity = LintingSeverity.Info;
+                            break;
+                        case HaxeDiagnosticsSeverity.ERROR:
+                            result.Severity = LintingSeverity.Error;
+                            break;
+                        case HaxeDiagnosticsSeverity.WARNING:
+                            result.Severity = LintingSeverity.Warning;
+                            break;
+                        default:
+                            continue;
                     }
 
-                    if (progress == total)
+                    switch (res.Kind)
                     {
-                        callback(list);
+                        case HaxeDiagnosticsKind.UnusedImport:
+                            result.Description = TextHelper.GetString("Info.UnusedImport");
+                            break;
+                        case HaxeDiagnosticsKind.UnresolvedIdentifier:
+                            result.Description = TextHelper.GetString("Info.UnresolvedIdentifier");
+                            break;
+                        case HaxeDiagnosticsKind.CompilerError:
+                        case HaxeDiagnosticsKind.RemovableCode:
+                            result.Description = res.Args.Description;
+                            break;
+                        default: //in case new kinds are added in new compiler versions
+                            continue;
                     }
+                    
+                    list.Add(result);
+                }
+            }
+            else if (status == HaxeCompleteStatus.ERROR)
+            {
+                PluginBase.RunAsync(() =>
+                {
+                    TraceManager.Add(hc.Errors, (int)TraceType.Error);
                 });
             }
         }
     }
+
+    class ProcessingQueue
+    {
+        readonly BlockingCollection<Action<Action>> queue = new BlockingCollection<Action<Action>>(new ConcurrentQueue<Action<Action>>());
+        readonly int maxRunning;
+        readonly HashSet<Task> running = new HashSet<Task>(); //TODO: running is modified on different threads
+
+        public ProcessingQueue(int maxConcurrent)
+        {
+            maxRunning = maxConcurrent;
+        }
+
+        public void Run(Action<Action> action)
+        {
+            lock (running)
+            {
+                if (running.Count < maxRunning)
+                {
+                    running.Add(CreateTask(action));
+                }
+                else
+                {
+                    queue.Add(action);
+                }
+            }
+        }
```

Comment:
Just some API design comment. Personally I think that if you would return the Task you can make this a bit more elegant and nicer. But no need to change anything.